### PR TITLE
feat: stack overflow detection exploiting reserved page

### PR DESCRIPTION
**Pull request by crusso**
Alternative to https://github.com/dfinity/motoko/pull/3782

* reserve last page of memory for Rust stack overflow detection
* instrument stack allocations that might not land on last page (but a previous page) to force an explicit access violation (by reading from last, unallocated word in full wasm memory.)
* omit bounds checks to accesses that will land within last page, trigger AV on subsequent access

Advantage over https://github.com/dfinity/motoko/pull/3782:

* most stack guards for static allocations are left implicit (less instructions, less cost).
* all stack overflows are reported as access violations (IC "heap out of bounds"), not some as "RTS Stack underflow" and some as "heap out of bounds"
* no need for hacky stack reservation at (logical) top of stack for Rust/C functions.

### CODE REVIEW

File: src/codegen/compile.ml [hunk 1]
```diff
@@ -1150,22 +1150,43 @@ module Stack = struct
...
 
   let get_stack_ptr env =
     G.i (GlobalGet (nr (E.get_global env "__stack_pointer")))
   let set_stack_ptr env =
     G.i (GlobalSet (nr (E.get_global env "__stack_pointer")))
 
-  (* TODO: check for overflow *)
+  let stack_overflow env =
+    Func.share_code0 env "stack_overflow" [] (fun env ->
+      (* read last byte of reserved page to force trap *)
```
Comment (crusso) [inline src/codegen/compile.ml:1172]:
```suggestion
      (* read last word of reserved page to force trap *)
```

File: src/codegen/compile.ml [hunk 2]
```diff
@@ -1150,22 +1150,43 @@ module Stack = struct
...
+    Func.share_code0 env "stack_overflow" [] (fun env ->
+      (* read last word of reserved page to force trap *)
+      compile_unboxed_const 0xFFFF_FFFCl ^^
+      G.i (Load {ty = I32Type; align = 2; offset = 0l; sz = None}) ^^
+      G.i Drop
+    )
+
   let alloc_words env n =
+    (* first, check for stack underflow, if necessary *)
+    (if (n >= page_size) then
```
Comment (crusso) [inline src/codegen/compile.ml:1180]:
```suggestion
    (if (n >= Int32.div(page_size, 4)) then
```

File: src/codegen/compile.ml [hunk 3]
```diff
@@ -1150,24 +1150,46 @@ module Stack = struct
...
+      (* read last word of reserved page to force trap *)
+      compile_unboxed_const 0xFFFF_FFFCl ^^
+      G.i (Load {ty = I32Type; align = 2; offset = 0l; sz = None}) ^^
+      G.i Drop
+    )
+
   let alloc_words env n =
+    let n_bytes = Int32.mul n Heap.word_size in
+    (* first, check for stack underflow, if necessary *)
+    (if (n_bytes >= page_size) then
```
Comment (ggreif) [inline src/codegen/compile.ml:1181]: Why not do this after line 1192, like
``` wasm
get_stack_ptr env
clz
if0
  G.nop
  (stack_overflow env)
```

File: src/codegen/compile.ml [hunk 4]
```diff
@@ -1150,24 +1150,46 @@ module Stack = struct
...
+      get_stack_ptr env ^^
+      compile_unboxed_const n_bytes ^^
+      G.i (Compare (Wasm.Values.I32 I32Op.LtU)) ^^
+      (G.if0
```
Comment (ggreif) [inline src/codegen/compile.ml:1185]: parens are redundant

File: src/codegen/compile.ml [hunk 5]
```diff
@@ -1150,24 +1150,46 @@ module Stack = struct
...
   let set_stack_ptr env =
     G.i (GlobalSet (nr (E.get_global env "__stack_pointer")))
 
-  (* TODO: check for overflow *)
+  let stack_overflow env =
+    Func.share_code0 env "stack_overflow" [] (fun env ->
+      (* read last word of reserved page to force trap *)
+      compile_unboxed_const 0xFFFF_FFFCl ^^
+      G.i (Load {ty = I32Type; align = 2; offset = 0l; sz = None}) ^^
+      G.i Drop
```
Comment (ggreif) [inline src/codegen/compile.ml:1175]: `G.Unreachable`?

File: test/run-drun/destabilization-crash/destabilization-crash.mo [hunk 6]
```diff
@@ -0,0 +1,42 @@
+import Prim = "mo:prim";
+
+// test desabilization of stabilized data, without rts stack overflow.
```
Comment (ggreif) [inline test/run-drun/destabilization-crash/destabilization-crash.mo:3]:
```suggestion
// test destabilization of stable variables, without rts stack overflow.
```

**Review by ggreif — APPROVED** [review]
Looks good, but I think we can shave off a few more instructions.

File: src/codegen/compile.ml [hunk 7]
```diff
@@ -1150,24 +1150,46 @@ module Stack = struct
...
+    )
+
   let alloc_words env n =
+    let n_bytes = Int32.mul n Heap.word_size in
```
Comment (ggreif) [inline src/codegen/compile.ml:1179]: assert that this is compatible with `Flags.rts_stack_pages`

**Comment by crusso** [comment]
> Looks good, but I think we can shave off a few more instructions.

Perhaps, but that can come in later PRs. I really want to make progress on getting parity for serialization and deserialization stack depth.

**Comment by ggreif** [comment]
> > Looks good, but I think we can shave off a few more instructions.
> 
> Perhaps, but that can come in later PRs. I really want to make progress on getting parity for serialization and deserialization stack depth.

I am following up in 3801.

**Pull request merged**